elements/pfe-avatar/src/pfe-avatar.js

```diff
@@ -0,0 +1,352 @@
+    this._canvas.height = size;
+
+    this._squareSize = this._canvas.width / 8;
+    this._triangleSize = this._canvas.width / 4;
+
+    this._ctx = this._canvas.getContext("2d");
+  }
+
+  static _registerColors() {
+    this.colors = [];
+    const themeColors =
+      PFElement.var("--pfe-avatar--colors") || this.defaultColors;
+
+    themeColors.split(/\s+/).forEach(colorCode => {
+      let pattern;
+      switch (colorCode.length) {
+        case 4: // ex: "#0fc"
+          pattern = /^#([A-f0-9])([A-f0-9])([A-f0-9])$/.exec(colorCode);
+          if (pattern) {
+            pattern.shift();
+            const color = pattern.map(c => parseInt(c + c, 16));
+            this._registerColor(color);
+          } else {
+            PFElement.log(`[pfe-avatar] invalid color ${colorCode}`);
+          }
+          break;
+        case 7: // ex: "#00ffcc"
+          pattern = /^#([A-f0-9]{2})([A-f0-9]{2})([A-f0-9]{2})$/.exec(
+            colorCode
+          );
+          if (pattern) {
+            pattern.shift();
+            const color = pattern.map(c => parseInt(c, 16));
+            this._registerColor(color);
+          } else {
+            PFElement.log(`[pfe-avatar] invalid color ${colorCode}`);
+          }
+      }
+    });
+
+    return this.colors;
+  }
+
+  static _registerColor(color) {
+    PfeAvatar.colors.push({
+      color1: `rgb(${color.join(",")})`,
+      color2: `rgb(${this._adjustColor(color).join(",")})`
+    });
+  }
+
+  static _adjustColor(color) {
+    const dark = 0.1;
+    const l_adj = 0.1; // luminance adjustment
+    const hsl = rgb2hsl(...color);
+
+    // if luminance is too dark already, then lighten the alternate color
+    // instead of darkening it.
+    hsl[2] += hsl[2] > dark ? -l_adj : l_adj;
+
+    return hsl2rgb(...hsl);
+  }
+
+  update() {
+    // if we have a src element, update the img, otherwise update the random pattern
+    if (this.hasAttribute("src")) {
+      this.shadowRoot.querySelector("img").src = this.src;
+    } else {
+      const bitPattern = hash(this.name).toString(2);
+      const arrPattern = bitPattern.split("").map(n => Number(n));
+      this._colorIndex = Math.floor(
+        (PfeAvatar.colors.length * parseInt(bitPattern, 2)) / Math.pow(2, 32)
+      );
+      this.color1 = PfeAvatar.colors[this._colorIndex].color1;
+      this.color2 = PfeAvatar.colors[this._colorIndex].color2;
+
+      this._clear();
+      this._drawBackground();
+      if (this.pattern === PfeAvatar.patterns.squares) {
+        this._drawSquarePattern(arrPattern);
+      } else if (this.pattern === PfeAvatar.patterns.triangles) {
+        this._drawTrianglePattern(arrPattern);
+      }
+      // this._drawGradient();
+    }
+  }
+
+  _clear() {
+    this._ctx.clearRect(0, 0, this._canvas.width, this._canvas.height);
+  }
+
+  _drawBackground() {
+    this._ctx.fillStyle = this.color1;
+    this._ctx.fillRect(0, 0, this._canvas.width, this._canvas.height);
+  }
+
+  _drawSquarePattern(pattern) {
```

Comment:
If you have time, can you add a short comment above each of these functions briefly describing what they do?

Reply:
Sure, I'll need to do that in a later sprint but definitely.  I have a (fairly old) branch which adds jsdoc-powered documentation which lines up nicely with this request.

Reply:
The issue for that branch is #89